contracts/javascore/centralized-connection/src/main/java/xcall/adapter/centralized/CentralizedConnection.java

```diff
@@ -0,0 +1,183 @@
+     */
+    @External
+    public void setAdmin(Address _relayer) {
+        OnlyAdmin();
+        adminAddress.set(_relayer);
+    }
+
+     /**
+      * Retrieves the admin address.
+      *
+      * @return The admin address.
+      */
+     @External(readonly = true)
+     public Address admin() {
+         return adminAddress.get();
+     }
+ 
+     /**
+      * Sets the fee to the target network
+      *
+      * @param  networkId      String Network Id of target chain
+      * @param  messageFee     The fee needed to send a Message 
+      * @param  responseFee    The fee of the response 
+      */
+     @External
+     public void setFee(String networkId, BigInteger messageFee, BigInteger responseFee) {
+        OnlyAdmin();
+        messageFees.set(networkId, messageFee);
+        responseFees.set(networkId, responseFee);
+     }
+ 
+     /**
+      * Returns the fee associated with the given destination address.
+      *
+      * @param  to       String Network Id of target chain
+      * @param  response whether the responding fee is included
+      * @return          The fee of sending a message to a given destination network
+      */
+     @External(readonly = true)
+     public BigInteger getFee(String to, boolean response) {
+         BigInteger messageFee = messageFees.get(to);
+         if (response) {
+             BigInteger responseFee = responseFees.get(to);
+             return messageFee.add(responseFee);
+         }
+         return messageFee;
+     }
+ 
+     /**
+      * Sends a message to the specified network.
+      *
+      * @param  to     Network Id of destination network 
+      * @param  svc    name of the service 
+      * @param  sn     positive for two-way message, zero for one-way message, negative for response
+      * @param  msg    serialized bytes of Service Message 
+      */
+     @Payable
+     @External
+     public void sendMessage(String to, String svc, BigInteger sn, byte[] msg) {
+         Context.require(Context.getCaller().equals(xCall.get()), "Only xCall can send messages");
+         BigInteger fee = BigInteger.ZERO;
+         if(sn.compareTo(BigInteger.ZERO)>=0) {
+             fee = getFee(to, true);
+         } else {
+            fee = getFee(to, false);
+         }
+
+         Context.require(Context.getValue().compareTo(fee) >= 0, "Insufficient balance");
+         Message(to, sn, msg);
+     }
+ 
+     /**
+      * Receives a message from a source network.
+      *
+      * @param  srcNetwork  the source network id from which the message is received
+      * @param  sn          the serial number of the message
+      * @param  msg         serialized bytes of Service Message 
+      */
+     @Payable
+     @External
+     public void recvMessage(String srcNetwork, BigInteger sn, byte[] msg) {
+        OnlyAdmin();
+        Context.require(!receipts.at(srcNetwork).getOrDefault(sn, false),"Duplicate Message");
+        receipts.at(srcNetwork).set(sn, true);
```

Comment:
same as in solidity contract, I believe we can not use sn for duplicate message.

Reply:
We should be able to, But i see that i missed this in the overall design pointers. 
SN included in the actual messages is never 0, if we use this when receiving messages we can avoid duplicates. Just requires a minimal computation on the relay side (For one way messages). But we should here also require it to not be 0. To make sure sn is always extracted.

Reply:
use conn sn